RELEASE_HOME is required even for assembleDebug

Only releaseRelease should trigger an error.

```
Caused by: java.lang.IllegalArgumentException: Please set RELEASE_HOME environment variable to a directory.
        at net.twisterrob.gradle.android.AndroidReleasePlugin$createReleaseTask$releaseVariantTask$1.invoke(AndroidReleasePlugin.kt:77)
        at net.twisterrob.gradle.android.AndroidReleasePlugin$createReleaseTask$releaseVariantTask$1.invoke(AndroidReleasePlugin.kt:24)
        at net.twisterrob.gradle.android.AndroidReleasePlugin$inlined$sam$i$org_gradle_api_Action$0.execute(PolymorphicDomainObjectContainerExtensions.kt)
        at org.gradle.api.internal.tasks.DefaultTaskContainer.create(DefaultTaskContainer.java:366)
        at org.gradle.api.internal.tasks.DefaultTaskContainer.create(DefaultTaskContainer.java:76)
        at net.twisterrob.gradle.android.AndroidReleasePlugin.createReleaseTask(AndroidReleasePlugin.kt:152)
        at net.twisterrob.gradle.android.AndroidReleasePlugin.access$createReleaseTask(AndroidReleasePlugin.kt:24)
        at net.twisterrob.gradle.android.AndroidReleasePlugin$createReleaseTasks$1.invoke(AndroidReleasePlugin.kt:51)
        at net.twisterrob.gradle.android.AndroidReleasePlugin$createReleaseTasks$3$1.invoke(AndroidReleasePlugin.kt:61)
        at net.twisterrob.gradle.android.AndroidReleasePlugin$createReleaseTasks$3$1.invoke(AndroidReleasePlugin.kt:24)
        at net.twisterrob.gradle.android.AndroidReleasePluginKt$sam$org_gradle_api_Action$0.execute(AndroidReleasePlugin.kt)
        at org.gradle.configuration.internal.DefaultUserCodeApplicationContext$CurrentApplication$1.execute(DefaultUserCodeApplicationContext.java:100)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:95)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction.execute(DefaultCollectionCallbackActionDecorator.java:92)
        at org.gradle.api.internal.collections.CollectionFilter$1.execute(CollectionFilter.java:59)
        at org.gradle.api.internal.DefaultDomainObjectCollection.all(DefaultDomainObjectCollection.java:163)
        at net.twisterrob.gradle.android.AndroidReleasePlugin$createReleaseTasks$3.invoke(AndroidReleasePlugin.kt:61)
        at net.twisterrob.gradle.android.AndroidReleasePlugin$createReleaseTasks$3.invoke(AndroidReleasePlugin.kt:24)
        at net.twisterrob.gradle.android.AndroidReleasePlugin$inlined$sam$i$org_gradle_api_Action$0.execute(PolymorphicDomainObjectContainerExtensions.kt)
        at org.gradle.configuration.internal.DefaultUserCodeApplicationContext$CurrentApplication$1.execute(DefaultUserCodeApplicationContext.java:100)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:95)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction.execute(DefaultCollectionCallbackActionDecorator.java:92)
        at org.gradle.api.internal.collections.CollectionFilter$1.execute(CollectionFilter.java:59)
        at org.gradle.api.internal.DefaultDomainObjectCollection.all(DefaultDomainObjectCollection.java:163)
        at org.gradle.api.internal.DefaultDomainObjectCollection.withType(DefaultDomainObjectCollection.java:205)
        at net.twisterrob.gradle.android.AndroidReleasePlugin.createReleaseTasks(AndroidReleasePlugin.kt:150)
        at net.twisterrob.gradle.android.AndroidReleasePlugin.access$createReleaseTasks(AndroidReleasePlugin.kt:24)
        at net.twisterrob.gradle.android.AndroidReleasePlugin$apply$1.execute(AndroidReleasePlugin.kt:39)
        at net.twisterrob.gradle.android.AndroidReleasePlugin$apply$1.execute(AndroidReleasePlugin.kt:24)
        at org.gradle.configuration.internal.DefaultUserCodeApplicationContext$CurrentApplication$1.execute(DefaultUserCodeApplicationContext.java:100)
        at org.gradle.configuration.internal.DefaultListenerBuildOperationDecorator$BuildOperationEmittingAction$1.run(DefaultListenerBuildOperationDecorator.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
        at org.gradle.configuration.internal.DefaultListenerBuildOperationDecorator$BuildOperationEmittingAction.execute(DefaultListenerBuildOperationDecorator.java:150)
        at org.gradle.internal.event.BroadcastDispatch$ActionInvocationHandler.dispatch(BroadcastDispatch.java:95)
        at org.gradle.internal.event.BroadcastDispatch$ActionInvocationHandler.dispatch(BroadcastDispatch.java:83)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:43)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:245)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:157)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:61)
        at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:346)
        at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:249)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:141)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
        at com.sun.proxy.$Proxy38.afterEvaluate(Unknown Source)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate$1.execute(LifecycleProjectEvaluator.java:183)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate$1.execute(LifecycleProjectEvaluator.java:180)
        at org.gradle.api.internal.project.DefaultProject.stepEvaluationListener(DefaultProject.java:1454)
        at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate.run(LifecycleProjectEvaluator.java:189)
        ... 194 more
```

--- a/plugin/plugin/build.gradle
+++ b/plugin/plugin/build.gradle
@@ -62,6 +62,8 @@ dependencies { // test
 	testImplementation("org.junit.jupiter:junit-jupiter-params:5.7.2")
 	testRuntimeOnly("org.junit.jupiter:junit-jupiter-engine:5.7.2")
 
+	testImplementation("org.junit-pioneer:junit-pioneer:1.4.2")
+
 	testImplementation("org.hamcrest:hamcrest:2.2") {
 		exclude group: "org.junit", module: "junit"
 	}

--- a/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
+++ b/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
@@ -12,10 +12,10 @@ import net.twisterrob.gradle.kotlin.dsl.extensions
 import org.gradle.api.DomainObjectSet
 import org.gradle.api.Project
 import org.gradle.api.Task
+import org.gradle.api.UnknownTaskException
 import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.StopExecutionException
 import org.gradle.api.tasks.bundling.Zip
-import org.gradle.kotlin.dsl.create
 import org.gradle.kotlin.dsl.getByName
 import org.gradle.kotlin.dsl.register
 import org.gradle.kotlin.dsl.withType
@@ -32,25 +32,25 @@ class AndroidReleasePlugin : BasePlugin() {
 
 		android = project.extensions.getByName<BaseExtension>("android")
 
-		val allTask = project.tasks.create<Task>("release") {
+		val allTask = project.tasks.register<Task>("release") {
 			group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
 			description = "Assembles and archives all builds"
 		}
 		project.afterEvaluate {
 			android.buildTypes.forEach { buildType ->
-				allTask.dependsOn(createReleaseTasks(buildType))
+				allTask.configure { it.dependsOn(createReleaseTasks(buildType)) }
 			}
 		}
 	}
 
-	private fun createReleaseTasks(buildType: BuildType): Task {
-		val allBuildTypeTask = project.tasks.create<Task>("releaseAll${buildType.name.capitalize()}") {
+	private fun createReleaseTasks(buildType: BuildType): Provider<out Task> {
+		val allBuildTypeTask = project.tasks.register<Task>("releaseAll${buildType.name.capitalize()}") {
 			group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
 			description = "Assembles and archives all ${buildType.name} builds"
 		}
 		LOG.debug("Creating tasks for {}", buildType.name)
 		fun createReleaseTaskWithDependency(variant: ApkVariant) {
-			allBuildTypeTask.dependsOn(createReleaseTask(variant))
+			allBuildTypeTask.configure { it.dependsOn(createReleaseTask(variant)) }
 		}
 
 		@Suppress("UNCHECKED_CAST")
@@ -121,14 +121,18 @@ class AndroidReleasePlugin : BasePlugin() {
 
 		variant.productFlavors.forEach { flavor ->
 			val releaseFlavorTaskName = "release${flavor.name.capitalize()}"
-			var releaseFlavorTask = project.tasks.findByName(releaseFlavorTaskName)
+			var releaseFlavorTask = try {
+				project.tasks.named(releaseFlavorTaskName)
+			} catch (ex: UnknownTaskException) {
+				null
+			}
 			if (releaseFlavorTask == null) {
-				releaseFlavorTask = project.tasks.create<Task>(releaseFlavorTaskName) {
+				releaseFlavorTask = project.tasks.register<Task>(releaseFlavorTaskName) {
 					group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
 					description = "Assembles and archives all builds for flavor ${flavor.name}"
 				}
 			}
-			releaseFlavorTask.dependsOn(releaseVariantTask)
+			releaseFlavorTask.configure { it.dependsOn(releaseVariantTask) }
 		}
 		return releaseVariantTask
 		/*val task: AndroidReleaseTask = variant.install.project.tasks.create(

--- a/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
+++ b/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
@@ -7,14 +7,16 @@ import com.android.build.gradle.LibraryPlugin
 import com.android.build.gradle.api.ApkVariant
 import com.android.build.gradle.internal.api.TestedVariant
 import com.android.builder.model.BuildType
+import com.android.builder.model.ProductFlavor
 import net.twisterrob.gradle.base.BasePlugin
 import net.twisterrob.gradle.kotlin.dsl.extensions
+import org.gradle.api.DomainObjectCollection
 import org.gradle.api.DomainObjectSet
 import org.gradle.api.Project
 import org.gradle.api.Task
 import org.gradle.api.UnknownTaskException
-import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.StopExecutionException
+import org.gradle.api.tasks.TaskProvider
 import org.gradle.api.tasks.bundling.Zip
 import org.gradle.kotlin.dsl.getByName
 import org.gradle.kotlin.dsl.invoke
@@ -44,31 +46,38 @@ class AndroidReleasePlugin : BasePlugin() {
 
 		android = project.extensions.getByName<BaseExtension>("android")
 
-		val allTask = project.tasks.register<Task>("release") {
-			group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
-			description = "Assembles and archives all builds"
-		}
+		val allTask = registerReleaseAllTask()
 		project.afterEvaluate {
 			android.buildTypes.forEach { buildType ->
-				val registerReleaseTasks = registerReleaseTasks(buildType)
-				allTask { dependsOn(registerReleaseTasks) }
+				val releaseBuildTypeTask = registerReleaseTasks(buildType)
+				allTask { dependsOn(releaseBuildTypeTask) }
 			}
 		}
 	}
 
-	private fun registerReleaseTasks(buildType: BuildType): Provider<out Task> {
+	private fun registerReleaseAllTask(): TaskProvider<Task> =
+		project.tasks.register<Task>("release") {
+			group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
+			description = "Assembles and archives all builds"
+		}
+
+	private fun registerReleaseTasks(buildType: BuildType): TaskProvider<Task> {
 		val allBuildTypeTask = project.tasks.register<Task>("releaseAll${buildType.name.capitalize()}") {
 			group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
 			description = "Assembles and archives all ${buildType.name} builds"
 		}
 		LOG.debug("Creating tasks for {}", buildType.name)
 		fun registerReleaseTaskWithDependency(variant: ApkVariant) {
-			val registerReleaseTask = registerReleaseTask(variant)
-			allBuildTypeTask { dependsOn(registerReleaseTask) }
+			val buildTypeTask = registerReleaseTask(variant)
+			allBuildTypeTask { dependsOn(buildTypeTask) }
+			variant.productFlavors.forEach { flavor ->
+				val flavorTask = registerFlavorTask(flavor)
+				flavorTask { dependsOn(buildTypeTask) }
+			}
 		}
 
 		@Suppress("UNCHECKED_CAST")
-		fun getVariantsForBuildType() =
+		fun getVariantsForBuildType(): DomainObjectCollection<ApkVariant> =
 			android.variants.matching { it.buildType.name == buildType.name } as DomainObjectSet<ApkVariant>
 
 		project.plugins.withType<AppPlugin> {
@@ -84,7 +93,7 @@ class AndroidReleasePlugin : BasePlugin() {
 		return allBuildTypeTask
 	}
 
-	private fun registerReleaseTask(variant: ApkVariant): Provider<out Task> {
+	private fun registerReleaseTask(variant: ApkVariant): TaskProvider<Zip> {
 		val releaseVariantTask = project.tasks.register<Zip>("release${variant.name.capitalize()}") {
 			group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
 			description = "Assembles and archives apk and its proguard mapping for ${variant.description}"
@@ -128,22 +137,6 @@ class AndroidReleasePlugin : BasePlugin() {
 		if (variant is TestedVariant && variant.testVariant != null) {
 			releaseVariantTask { dependsOn(variant.testVariant.assembleProvider) }
 		}
-
-		variant.productFlavors.forEach { flavor ->
-			val releaseFlavorTaskName = "release${flavor.name.capitalize()}"
-			var releaseFlavorTask = try {
-				project.tasks.named(releaseFlavorTaskName)
-			} catch (ex: UnknownTaskException) {
-				null
-			}
-			if (releaseFlavorTask == null) {
-				releaseFlavorTask = project.tasks.register<Task>(releaseFlavorTaskName) {
-					group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
-					description = "Assembles and archives all builds for flavor ${flavor.name}"
-				}
-			}
-			releaseFlavorTask { dependsOn(releaseVariantTask) }
-		}
 		return releaseVariantTask
 		/*val task: AndroidReleaseTask = variant.install.project.tasks.create(
 			mapOf(
@@ -153,5 +146,22 @@ class AndroidReleasePlugin : BasePlugin() {
 			)
 		)
 		task.variant = variant*/
+	}
+
+	private fun registerFlavorTask(flavor: ProductFlavor): TaskProvider<Task> {
+		val releaseFlavorTaskName = "release${flavor.name.capitalize()}"
+		// Get the flavor task in case it was already registered by another variant.
+		var releaseFlavorTask = try {
+			project.tasks.named(releaseFlavorTaskName)
+		} catch (ex: UnknownTaskException) {
+			null
+		}
+		if (releaseFlavorTask == null) {
+			releaseFlavorTask = project.tasks.register<Task>(releaseFlavorTaskName) {
+				group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
+				description = "Assembles and archives all builds for flavor ${flavor.name}"
+			}
+		}
+		return releaseFlavorTask
 	}
 }

--- a/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
+++ b/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
@@ -17,6 +17,7 @@ import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.StopExecutionException
 import org.gradle.api.tasks.bundling.Zip
 import org.gradle.kotlin.dsl.getByName
+import org.gradle.kotlin.dsl.invoke
 import org.gradle.kotlin.dsl.register
 import org.gradle.kotlin.dsl.withType
 import java.io.File
@@ -38,7 +39,8 @@ class AndroidReleasePlugin : BasePlugin() {
 		}
 		project.afterEvaluate {
 			android.buildTypes.forEach { buildType ->
-				allTask.configure { it.dependsOn(registerReleaseTasks(buildType)) }
+				val registerReleaseTasks = registerReleaseTasks(buildType)
+				allTask { dependsOn(registerReleaseTasks) }
 			}
 		}
 	}
@@ -50,7 +52,8 @@ class AndroidReleasePlugin : BasePlugin() {
 		}
 		LOG.debug("Creating tasks for {}", buildType.name)
 		fun registerReleaseTaskWithDependency(variant: ApkVariant) {
-			allBuildTypeTask.configure { it.dependsOn(registerReleaseTask(variant)) }
+			val registerReleaseTask = registerReleaseTask(variant)
+			allBuildTypeTask { dependsOn(registerReleaseTask) }
 		}
 
 		@Suppress("UNCHECKED_CAST")
@@ -114,9 +117,9 @@ class AndroidReleasePlugin : BasePlugin() {
 				println("Published release artifacts to ${outputs.files.singleFile}")
 			}
 		}
-		releaseVariantTask.configure { it.dependsOn(variant.assembleProvider) }
+		releaseVariantTask { dependsOn(variant.assembleProvider) }
 		if (variant is TestedVariant && variant.testVariant != null) {
-			releaseVariantTask.configure { it.dependsOn(variant.testVariant.assembleProvider) }
+			releaseVariantTask { dependsOn(variant.testVariant.assembleProvider) }
 		}
 
 		variant.productFlavors.forEach { flavor ->
@@ -132,7 +135,7 @@ class AndroidReleasePlugin : BasePlugin() {
 					description = "Assembles and archives all builds for flavor ${flavor.name}"
 				}
 			}
-			releaseFlavorTask.configure { it.dependsOn(releaseVariantTask) }
+			releaseFlavorTask { dependsOn(releaseVariantTask) }
 		}
 		return releaseVariantTask
 		/*val task: AndroidReleaseTask = variant.install.project.tasks.create(

--- a/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
+++ b/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
@@ -12,10 +12,12 @@ import net.twisterrob.gradle.kotlin.dsl.extensions
 import org.gradle.api.DomainObjectSet
 import org.gradle.api.Project
 import org.gradle.api.Task
+import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.StopExecutionException
 import org.gradle.api.tasks.bundling.Zip
 import org.gradle.kotlin.dsl.create
 import org.gradle.kotlin.dsl.getByName
+import org.gradle.kotlin.dsl.register
 import org.gradle.kotlin.dsl.withType
 import java.io.File
 
@@ -68,8 +70,8 @@ class AndroidReleasePlugin : BasePlugin() {
 		return allBuildTypeTask
 	}
 
-	private fun createReleaseTask(variant: ApkVariant): Task {
-		val releaseVariantTask = project.tasks.create<Zip>("release${variant.name.capitalize()}") {
+	private fun createReleaseTask(variant: ApkVariant): Provider<out Task> {
+		val releaseVariantTask = project.tasks.register<Zip>("release${variant.name.capitalize()}") {
 			group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
 			description = "Assembles and archives apk and its proguard mapping for ${variant.description}"
 			val releaseDir = File(
@@ -112,9 +114,9 @@ class AndroidReleasePlugin : BasePlugin() {
 				println("Published release artifacts to ${outputs.files.singleFile}")
 			}
 		}
-		releaseVariantTask.dependsOn(variant.assembleProvider)
+		releaseVariantTask.configure { it.dependsOn(variant.assembleProvider) }
 		if (variant is TestedVariant && variant.testVariant != null) {
-			releaseVariantTask.dependsOn(variant.testVariant.assembleProvider)
+			releaseVariantTask.configure { it.dependsOn(variant.testVariant.assembleProvider) }
 		}
 
 		variant.productFlavors.forEach { flavor ->

--- a/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
+++ b/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
@@ -38,19 +38,19 @@ class AndroidReleasePlugin : BasePlugin() {
 		}
 		project.afterEvaluate {
 			android.buildTypes.forEach { buildType ->
-				allTask.configure { it.dependsOn(createReleaseTasks(buildType)) }
+				allTask.configure { it.dependsOn(registerReleaseTasks(buildType)) }
 			}
 		}
 	}
 
-	private fun createReleaseTasks(buildType: BuildType): Provider<out Task> {
+	private fun registerReleaseTasks(buildType: BuildType): Provider<out Task> {
 		val allBuildTypeTask = project.tasks.register<Task>("releaseAll${buildType.name.capitalize()}") {
 			group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
 			description = "Assembles and archives all ${buildType.name} builds"
 		}
 		LOG.debug("Creating tasks for {}", buildType.name)
-		fun createReleaseTaskWithDependency(variant: ApkVariant) {
-			allBuildTypeTask.configure { it.dependsOn(createReleaseTask(variant)) }
+		fun registerReleaseTaskWithDependency(variant: ApkVariant) {
+			allBuildTypeTask.configure { it.dependsOn(registerReleaseTask(variant)) }
 		}
 
 		@Suppress("UNCHECKED_CAST")
@@ -60,17 +60,17 @@ class AndroidReleasePlugin : BasePlugin() {
 		project.plugins.withType<AppPlugin> {
 			val matching = getVariantsForBuildType()
 			LOG.debug("Found android app, variants with {}: {}", buildType.name, matching)
-			matching.all(::createReleaseTaskWithDependency)
+			matching.all(::registerReleaseTaskWithDependency)
 		}
 		project.plugins.withType<LibraryPlugin> {
 			val matching = getVariantsForBuildType()
 			LOG.debug("Found android lib, variants with {}: {}", buildType.name, matching)
-			matching.all(::createReleaseTaskWithDependency)
+			matching.all(::registerReleaseTaskWithDependency)
 		}
 		return allBuildTypeTask
 	}
 
-	private fun createReleaseTask(variant: ApkVariant): Provider<out Task> {
+	private fun registerReleaseTask(variant: ApkVariant): Provider<out Task> {
 		val releaseVariantTask = project.tasks.register<Zip>("release${variant.name.capitalize()}") {
 			group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
 			description = "Assembles and archives apk and its proguard mapping for ${variant.description}"

--- a/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
+++ b/plugin/plugin/src/main/kotlin/net/twisterrob/gradle/android/AndroidReleasePlugin.kt
@@ -22,7 +22,18 @@ import org.gradle.kotlin.dsl.register
 import org.gradle.kotlin.dsl.withType
 import java.io.File
 
-private const val envVarName = "RELEASE_HOME"
+private val defaultReleaseDir: File
+	get() {
+		val envVarName = "RELEASE_HOME"
+		val releaseHome = checkNotNull(System.getenv(envVarName)) {
+			"Please set ${envVarName} environment variable to an existing directory."
+		}
+		return File(releaseHome).also {
+			check(it.exists() && it.isDirectory) {
+				"Please set ${envVarName} environment variable to an existing directory."
+			}
+		}
+	}
 
 class AndroidReleasePlugin : BasePlugin() {
 
@@ -77,11 +88,7 @@ class AndroidReleasePlugin : BasePlugin() {
 		val releaseVariantTask = project.tasks.register<Zip>("release${variant.name.capitalize()}") {
 			group = org.gradle.api.plugins.BasePlugin.UPLOAD_GROUP
 			description = "Assembles and archives apk and its proguard mapping for ${variant.description}"
-			val releaseDir = File(
-				System.getenv(envVarName)
-					?: throw IllegalArgumentException("Please set ${envVarName} environment variable to a directory.")
-			)
-			destinationDirectory.set(releaseDir.resolve("android"))
+			destinationDirectory.fileProvider(project.provider { defaultReleaseDir.resolve("android") })
 			archiveFileName.set(
 				android.defaultConfig
 					.extensions.getByName<AndroidVersionExtension>(AndroidVersionExtension.NAME)

--- a/plugin/plugin/src/test/kotlin/net/twisterrob/gradle/android/AndroidReleasePluginIntgTest.kt
+++ b/plugin/plugin/src/test/kotlin/net/twisterrob/gradle/android/AndroidReleasePluginIntgTest.kt
@@ -3,6 +3,7 @@ package net.twisterrob.gradle.android
 import net.twisterrob.gradle.test.GradleRunnerRule
 import net.twisterrob.gradle.test.GradleRunnerRuleExtension
 import net.twisterrob.gradle.test.assertHasOutputLine
+import net.twisterrob.gradle.test.assertNoTask
 import net.twisterrob.gradle.test.assertSuccess
 import net.twisterrob.gradle.test.root
 import net.twisterrob.test.zip.hasZipEntry
@@ -13,6 +14,7 @@ import org.hamcrest.io.FileMatchers.anExistingFile
 import org.intellij.lang.annotations.Language
 import org.junit.jupiter.api.Test
 import org.junit.jupiter.api.extension.ExtendWith
+import org.junitpioneer.jupiter.ClearEnvironmentVariable
 import java.io.File
 import java.time.Instant
 import java.util.zip.ZipFile
@@ -37,6 +39,18 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 				tasks.releaseDebug.destinationDirectory.set(file('releases'))
 			}
 		""".trimIndent()
+	}
+
+	@ClearEnvironmentVariable(key = "RELEASE_HOME")
+	@Test fun `assemble doesn't need env var`() {
+		val result = gradle.run(script, "assemble", "assembleAndroidTest").build()
+
+		result.assertSuccess(":assembleDebug")
+		result.assertSuccess(":assembleRelease")
+		result.assertSuccess(":assembleDebugAndroidTest")
+		result.assertNoTask(":releaseDebug")
+		result.assertNoTask(":releaseRelease")
+		result.assertNoTask(":releaseAll")
 	}
 
 	@Test fun `test (release)`() {

--- a/plugin/plugin/src/test/kotlin/net/twisterrob/gradle/android/AndroidReleasePluginIntgTest.kt
+++ b/plugin/plugin/src/test/kotlin/net/twisterrob/gradle/android/AndroidReleasePluginIntgTest.kt
@@ -27,22 +27,13 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 
 	override lateinit var gradle: GradleRunnerRule
 
-	companion object {
-
+	@ClearEnvironmentVariable(key = "RELEASE_HOME")
+	@Test fun `assemble doesn't need env var`() {
 		@Language("gradle")
 		val script = """
 			apply plugin: 'net.twisterrob.android-app'
-			android.defaultConfig.version { major = 1; minor = 2; patch = 3; build = 4 }
-			afterEvaluate {
-				// TODO workaround for testing until a Task input property is introduced instead of env.RELEASE_HOME
-				tasks.named("releaseRelease") { destinationDirectory.set(file('releases')) }
-				tasks.named("releaseDebug") { destinationDirectory.set(file('releases')) }
-			}
 		""".trimIndent()
-	}
 
-	@ClearEnvironmentVariable(key = "RELEASE_HOME")
-	@Test fun `assemble doesn't need env var`() {
 		val result = gradle.run(script, "assemble", "assembleAndroidTest").build()
 
 		result.assertSuccess(":assembleDebug")
@@ -53,7 +44,40 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 		result.assertNoTask(":releaseAll")
 	}
 
+	@ClearEnvironmentVariable(key = "RELEASE_HOME")
+	@Test fun `needs environment variable to release (debug)`() {
+		@Language("gradle")
+		val script = """
+			apply plugin: 'net.twisterrob.android-app'
+		""".trimIndent()
+
+		val result = gradle.run(script, "releaseDebug").buildAndFail()
+
+		result.assertHasOutputLine(""".*Please set RELEASE_HOME environment variable to an existing directory\.""".toRegex())
+	}
+
+	@ClearEnvironmentVariable(key = "RELEASE_HOME")
+	@Test fun `needs environment variable to release (release)`() {
+		@Language("gradle")
+		val script = """
+			apply plugin: 'net.twisterrob.android-app'
+		""".trimIndent()
+
+		val result = gradle.run(script, "releaseRelease").buildAndFail()
+
+		result.assertHasOutputLine(""".*Please set RELEASE_HOME environment variable to an existing directory\.""".toRegex())
+	}
+
 	@Test fun `test (release)`() {
+		@Language("gradle")
+		val script = """
+			apply plugin: 'net.twisterrob.android-app'
+			android.defaultConfig.version { major = 1; minor = 2; patch = 3; build = 4 }
+			afterEvaluate {
+				tasks.named("releaseRelease", Zip) { destinationDirectory.set(file('releases/release')) }
+			}
+		""".trimIndent()
+
 		val result = gradle.run(script, "releaseRelease").build()
 
 		result.assertSuccess(":assembleRelease")
@@ -62,7 +86,7 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 	}
 
 	private fun assertReleaseArchive(result: BuildResult) {
-		val releasesDir = gradle.root.resolve("releases")
+		val releasesDir = gradle.root.resolve("releases/release")
 		assertArchive(releasesDir.resolve("${packageName}@10203004-v1.2.3#4+archive.zip")) { archive ->
 			assertThat(archive, hasZipEntry("${packageName}@10203004-v1.2.3#4+release.apk"))
 			assertThat(archive, hasZipEntry("proguard_configuration.pro"))
@@ -78,6 +102,15 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 	}
 
 	@Test fun `test (debug)`() {
+		@Language("gradle")
+		val script = """
+			apply plugin: 'net.twisterrob.android-app'
+			android.defaultConfig.version { major = 1; minor = 2; patch = 3; build = 4 }
+			afterEvaluate {
+				tasks.named("releaseDebug", Zip) { destinationDirectory.set(file('releases/debug')) }
+			}
+		""".trimIndent()
+
 		val result = gradle.run(script, "releaseDebug").build()
 
 		result.assertSuccess(":assembleDebug")
@@ -87,7 +120,7 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 	}
 
 	private fun assertDebugArchive(result: BuildResult) {
-		val releasesDir = gradle.root.resolve("releases")
+		val releasesDir = gradle.root.resolve("releases/debug")
 		assertArchive(releasesDir.resolve("${packageName}.debug@10203004-v1.2.3#4d+archive.zip")) { archive ->
 			assertThat(archive, hasZipEntry("${packageName}.debug@10203004-v1.2.3#4d+debug.apk"))
 			assertThat(archive, hasZipEntry("${packageName}.debug.test@10203004-v1.2.3#4d+debug-androidTest.apk"))
@@ -104,6 +137,16 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 	}
 
 	@Test fun `test (debug) and (release)`() {
+		@Language("gradle")
+		val script = """
+			apply plugin: 'net.twisterrob.android-app'
+			android.defaultConfig.version { major = 1; minor = 2; patch = 3; build = 4 }
+			afterEvaluate {
+				tasks.named("releaseRelease", Zip) { destinationDirectory.set(file('releases/release')) }
+				tasks.named("releaseDebug", Zip) { destinationDirectory.set(file('releases/debug')) }
+			}
+		""".trimIndent()
+
 		val result = gradle.run(script, "release").build()
 
 		result.assertSuccess(":assembleRelease")

--- a/plugin/plugin/src/test/kotlin/net/twisterrob/gradle/android/AndroidReleasePluginIntgTest.kt
+++ b/plugin/plugin/src/test/kotlin/net/twisterrob/gradle/android/AndroidReleasePluginIntgTest.kt
@@ -64,7 +64,7 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 	}
 
 	@Test fun `test (debug)`() {
-		val result = gradle.run(script, "releaseDebug", "--warning-mode", "all").build()
+		val result = gradle.run(script, "releaseDebug").build()
 
 		result.assertSuccess(":assembleDebug")
 		result.assertSuccess(":assembleDebugAndroidTest")

--- a/plugin/plugin/src/test/kotlin/net/twisterrob/gradle/android/AndroidReleasePluginIntgTest.kt
+++ b/plugin/plugin/src/test/kotlin/net/twisterrob/gradle/android/AndroidReleasePluginIntgTest.kt
@@ -35,8 +35,8 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 			android.defaultConfig.version { major = 1; minor = 2; patch = 3; build = 4 }
 			afterEvaluate {
 				// TODO workaround for testing until a Task input property is introduced instead of env.RELEASE_HOME
-				tasks.releaseRelease.destinationDirectory.set(file('releases'))
-				tasks.releaseDebug.destinationDirectory.set(file('releases'))
+				tasks.named("releaseRelease") { destinationDirectory.set(file('releases')) }
+				tasks.named("releaseDebug") { destinationDirectory.set(file('releases')) }
 			}
 		""".trimIndent()
 	}

--- a/plugin/plugin/src/test/kotlin/net/twisterrob/gradle/android/AndroidReleasePluginIntgTest.kt
+++ b/plugin/plugin/src/test/kotlin/net/twisterrob/gradle/android/AndroidReleasePluginIntgTest.kt
@@ -23,11 +23,11 @@ import java.util.zip.ZipFile
  * @see AndroidReleasePlugin
  */
 @ExtendWith(GradleRunnerRuleExtension::class)
+@ClearEnvironmentVariable(key = "RELEASE_HOME")
 class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 
 	override lateinit var gradle: GradleRunnerRule
 
-	@ClearEnvironmentVariable(key = "RELEASE_HOME")
 	@Test fun `assemble doesn't need env var`() {
 		@Language("gradle")
 		val script = """
@@ -44,7 +44,6 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 		result.assertNoTask(":releaseAll")
 	}
 
-	@ClearEnvironmentVariable(key = "RELEASE_HOME")
 	@Test fun `needs environment variable to release (debug)`() {
 		@Language("gradle")
 		val script = """
@@ -56,7 +55,6 @@ class AndroidReleasePluginIntgTest : BaseAndroidIntgTest() {
 		result.assertHasOutputLine(""".*Please set RELEASE_HOME environment variable to an existing directory\.""".toRegex())
 	}
 
-	@ClearEnvironmentVariable(key = "RELEASE_HOME")
 	@Test fun `needs environment variable to release (release)`() {
 		@Language("gradle")
 		val script = """

